src/main/java/org/mockbukkit/mockbukkit/inventory/meta/BlockStateMetaMock.java

```diff
@@ -0,0 +1,323 @@
+			}
+			else
+			{
+				throw new UnsupportedOperationException("TileStateMock for '" + material + "' has not been implemented by MockBukkit.");
+			}
+		}
+		catch (ReflectiveOperationException e)
+		{
+			throw new UnsupportedOperationException("Can't instantiate class '" + clazz + "'");
+		}
+	}
+
+	@Override
+	public void setBlockState(@NotNull BlockState blockState)
+	{
+		this.blockState = blockState;
+		if (this.material == null)
+		{
+			this.material = blockState.getType();
+		}
+	}
+
+	@Override
+	protected String getTypeName()
+	{
+		return "TILE_ENTITY";
+	}
+
+	@Override
+	public boolean equals(Object o)
+	{
+		if (this == o) return true;
+		if (!(o instanceof BlockStateMetaMock that)) return false;
+		if (!super.equals(o)) return false;
+		return Objects.equals(blockState, that.blockState) && material == that.material;
+	}
+
+	@Override
+	public int hashCode()
+	{
+		return Objects.hash(super.hashCode(), blockState, material);
+	}
+
+	@Override
+	public @NotNull BlockStateMetaMock clone()
+	{
+		BlockStateMetaMock clone = (BlockStateMetaMock) super.clone();
+		clone.blockState = this.blockState != null ? this.blockState.copy() : null;
+		return clone;
+	}
+
+	@Override
+	public @NotNull Map<String, Object> serialize()
+	{
+		Map<String, Object> serialized = super.serialize();
+
+		if (material != null)
+		{
+			serialized.put("blockMaterial", material.name());
+		}
+
+		if (blockState instanceof Container container)
+		{
+			ItemStack[] contents = container.getInventory().getContents();
+			List<Map<String, Object>> containerData = new ArrayList<>(contents.length);
+			for (int i = 0; i < contents.length; i++)
+			{
+				ItemStack item = contents[i];
+				if (item != null && item.getType() != Material.AIR)
+				{
+					containerData.add(Map.of("slot", i, "item", item.serialize()));
+				}
+			}
+			serialized.put("container", containerData);
+		}
+		// TODO: serialize other TileStates
+		return serialized;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	protected void deserializeInternal(@NotNull Map<String, Object> args)
+	{
+		clearBlockState();
+		super.deserializeInternal(args);
+
+		if (args.containsKey("blockMaterial"))
+		{
+			material = Registry.MATERIAL.get(NamespacedKey.fromString(((String) args.get("blockMaterial")).toLowerCase(Locale.ENGLISH)));
+		}
+		if (args.containsKey("container"))
+		{
+			if (material != null)
+			{
+				blockState = getBlockState();
+			}
+			if (blockState instanceof ContainerStateMock container)
+			{
+				Inventory inventory = container.getInventory();
+				List<Map<String, Object>> containerData = (List<Map<String, Object>>) args.get("container");
+				for (Map<String, Object> slotData : containerData)
+				{
+					int slot = (int) slotData.getOrDefault("slot", -1);
+					if (slot >= 0)
+					{
+						inventory.setItem(slot, ItemStack.deserialize((Map<String, Object>) slotData.get("item")));
+					}
+				}
+			}
+			// TODO: deserialize other TileStates
+		}
```

Comment:
Probably a good idea to unwrap this, such that it's easier to read. Like this, for say
```suggestion
		if (!args.containsKey("container"))
		{
			return;
		}
		if (material != null)
		{
			blockState = getBlockState();
		}
		if (!(blockState instanceof ContainerStateMock container))
		{
			// TODO: deserialize other TileStates (Other function call probably)
			return;
		}
		Inventory inventory = container.getInventory();
		List<Map<String, Object>> containerData = (List<Map<String, Object>>) args.get("container");
		for (Map<String, Object> slotData : containerData)
		{
			int slot = (int) slotData.getOrDefault("slot", -1);
			if (slot >= 0)
			{
				inventory.setItem(slot, ItemStack.deserialize((Map<String, Object>) slotData.get("item")));
			}
		}
```
I'd say that's more readable, at least for the moment. Probably a good idea to make a separate function for each different state